File: problem_19.cpp
#include <chrono>
#include <iostream>
#include <cmath>
#include <cstring>
#include <array>

static const int days_of_week = 7;
static const int months_of_year = 12;
static const std::array<int, months_of_year> months = {
    31,
    28,
    31,
    30,
    31,
    30,
    31,
    31,
    30,
    31,
    30,
    31
};

struct Date 
{
    int day;
    int month;
    int year;
    int dow;

    Date(int d, int m, int y, int w) 
    {
        day = d;
        month = m;
        year = y;
        dow = w;
    }

    Date(int d, int m, int y) 
    {
        day = d;
        month = m;
        year = y;
    }

    bool isLeapYear() 
    {
        return 
            year % 4 == 0
            && (year % 100 != 0 
            || year % 400 == 0);
    }

    int getMonthLen() 
    {
        if (!isLeapYear() || month != 2) 
        {
            return months.at(month - 1);
        }
        else 
        {
            return months.at(month - 1) + 1;
        }
    }

    void sumInterval(int days)
    {
        dow += days % days_of_week;
        if (day + days <= getMonthLen()) 
        {
            day += days;
            return;
        }
        else 
        {
            day = ((day + days) % (getMonthLen() + 1)) + 1;
            month += 1;
            if (month > months_of_year) 
            {
                month = 1;
                year += 1;
            }
        }
    }

    bool isLessThan(Date d)
    {
        if (year < d.year) 
        {
            return true;
        }
        else if (year > d.year) 
        {
            return false;
        }
        if (month < d.month) 
        {
            return true;
        }
        else if (month > d.month) 
        {
            return false;
        }
        if (day < d.day) 
        {
            return true;
        }
        return false;
    }
};

int main() 
{
    Date start(1, 1, 1900, 1);
    Date startCount(31, 12, 1900);
    Date endCount(1, 1, 2001);
    auto startTm = std::chrono::high_resolution_clock::now();

    bool startedCount = false;
    int countedSundays = 0;
    start.sumInterval(6); // move to first Sunday
    while (start.isLessThan(endCount)) 
    {
        if (startedCount || startCount.isLessThan(start))
        {
            startedCount = true;
            if (start.day == 1) 
            {
                countedSundays++;
            }
        }
        start.sumInterval(7); // move weekly
    }

    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << countedSundays << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_19.rs
use std::time::{Instant};

const DAYS_OF_WEEK: u8 = 7;
const MONTHS_OF_YEAR: usize = 12;
const MONTHS: [u8; MONTHS_OF_YEAR] = [
    31,
    28,
    31,
    30,
    31,
    30,
    31,
    31,
    30,
    31,
    30,
    31
];

struct Date 
{
    day: u8,
    month: u8,
    year: u16,
    dow: u8,
}

impl Date
{
    fn init (d: u8, m: u8, y: u16, w : Option<u8>) -> Date
    {
        return Date { day: d, month: m, year: y, dow: w.unwrap_or(0) };
    }

    fn is_leap_year(&self) -> bool
    {
        return 
            self.year % 4 == 0
            && (self.year % 100 != 0 
            || self.year % 400 == 0);
    }

    fn get_month_len(&self) -> u8
    {
        if !self.is_leap_year() || self.month != 2 
        {
            return MONTHS[(self.month - 1) as usize];
        }
        else 
        {
            return MONTHS[(self.month - 1) as usize] + 1;
        }
    }

    fn sum_interval(&mut self, days: u8)
    {
        self.dow += days % DAYS_OF_WEEK;
        if self.day + days <= self.get_month_len() 
        {
            self.day = self.day + days;
            return;
        }
        else 
        {
            self.day = ((self.day + days) % (self.get_month_len() + 1)) + 1;
            self.month += 1;
            if self.month > (MONTHS_OF_YEAR) as u8
            {
                self.month = 1;
                self.year += 1;
            }
        }
    }

    fn is_less_than(&self, d: &Date) -> bool
    {
        if self.year < d.year 
        {
            return true;
        }
        else if self.year > d.year 
        {
            return false;
        }
        if self.month < d.month 
        {
            return true;
        }
        else if self.month > d.month 
        {
            return false;
        }
        if self.day < d.day 
        {
            return true;
        }
        return false;
    }
}


fn main () 
{
    let start_tm = Instant::now();

    let mut start: Date = Date::init(1, 1, 1900, Some(1));
    let start_count: Date = Date::init(31, 12, 1900, None);
    let end_count: Date = Date::init(1, 1, 2001, None);

    let mut started_count: bool = false;
    let mut counted_sundays: u16 = 0;
    start.sum_interval(6); // move to first Sunday
    while start.is_less_than(&end_count) 
    {
        if started_count || start_count.is_less_than(&start)
        {
            started_count = true;
            if start.day == 1 
            {
                counted_sundays = counted_sundays + 1;
            }
        }
        start.sum_interval(7); // move weekly
    }

    println!("Result: {}", counted_sundays);
    let duration = start_tm.elapsed();
    println!("Timing: {}", duration.as_nanos());
}







